File: app/src/main/java/com/guillermonegrete/tts/textprocessing/domain/interactors/GetExternalLink.java
package com.guillermonegrete.tts.textprocessing.domain.interactors;

import com.guillermonegrete.tts.AbstractInteractor;
import com.guillermonegrete.tts.Executor;
import com.guillermonegrete.tts.MainThread;
import com.guillermonegrete.tts.data.source.ExternalLinksDataSource;
import com.guillermonegrete.tts.db.ExternalLink;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class GetExternalLink extends AbstractInteractor implements GetExternalLinksInteractor {

    private Callback callback;

    private ExternalLinksDataSource dataSource;

    private String language;

    public GetExternalLink(Executor executor, MainThread mainThread, Callback callback, ExternalLinksDataSource dataSource, String language) {
        super(executor, mainThread);
        this.callback = callback;
        this.dataSource = dataSource;
        this.language = language;
    }

    @Override
    public void run() {
        dataSource.getLanguageLinks(language, new ExternalLinksDataSource.Callback() {
            @Override
            public void onLinksRetrieved(@NotNull final List<? extends ExternalLink> links) {
                mMainThread.post(new Runnable() {
                    @Override
                    public void run() {
                        callback.onExternalLinksRetrieved((List<ExternalLink>) links);
                    }
                });
            }
        });
    }
}

File: app/src/main/java/com/guillermonegrete/tts/textprocessing/domain/interactors/GetExternalLink.kt
package com.guillermonegrete.tts.textprocessing.domain.interactors

import com.guillermonegrete.tts.AbstractInteractor
import com.guillermonegrete.tts.MainThread
import com.guillermonegrete.tts.data.source.ExternalLinksDataSource
import com.guillermonegrete.tts.db.ExternalLink
import java.util.concurrent.ExecutorService

class GetExternalLink(
    executor: ExecutorService,
    mainThread: MainThread,
    private val dataSource: ExternalLinksDataSource
) : AbstractInteractor(executor, mainThread), GetExternalLinksInteractor {

    private var language = ""
    private var callback: GetExternalLinksInteractor.Callback? = null

    override fun run() {
        dataSource.getLanguageLinks(language, object : ExternalLinksDataSource.Callback {
            override fun onLinksRetrieved(links: List<ExternalLink>) {
                mMainThread.post { callback?.onExternalLinksRetrieved(links) }
            }
        })
    }

    operator fun invoke(language: String, callback: GetExternalLinksInteractor.Callback){
        this.language = language
        this.callback = callback
        if(!executorService.isShutdown) executorService.execute { run() }
    }
}
